dashboard/app/controllers/datablock_storage_controller.rb

```diff
@@ -0,0 +1,347 @@
+# datablockStorage.js. Each method in this controller has a corresponding method there.
+#
+# Methods in this controller are available as (see `:datablock_storage` in routes.rb):
+# /datablock_storage/:channel_id/:method_name
+#
+# Student data is stored in MySQL in corresponding ActiveRecord models:
+# - datablock_storage_table.rb: stores a list of tables and their columns, after this file
+#   this model contains the most code.
+# - datablock_storage_record.rb: stores the records in each table
+# - datablock_storage_kvp.rb: stores key-value pairs
+#
+# Metadata for the code.org defined datasets are stored in:
+# - datablock_storage_library_manifest.rb
+#
+# Methods are broken into sections, for example the Key-Value-Pair API, Table API,
+# Table Column API, Table Record API, Library Manifest API & Project API
+#
+# More details can be found in the PR that initially created Datablock Storage:
+# https://github.com/code-dot-org/code-dot-org/pull/56279
+
+class DatablockStorageController < ApplicationController
+  before_action :validate_channel_id
+  before_action :authenticate_user!
+  skip_before_action :verify_authenticity_token
+
+  StudentFacingError = DatablockStorageTable::StudentFacingError
+
+  # A StudentFacingError can be thrown by any of the Datablock Storage models
+  # and indicates that the error should show up in the Applab "Debug Console".
+  #
+  # Here we catch any of these errors, and return it in the expected JSON format to
+  # datablockStorage.js, who in turn calls its onError() callback with appropriate data.
+  rescue_from StudentFacingError do |exception|
+    render json: {msg: exception.message, type: exception.type}, status: :bad_request
+  end
+
+  # We only permit requests to /datablock_storage/:channel_id for projects with
+  # the appropriate project_type (aka game.app).
+  SUPPORTED_PROJECT_TYPES = [
+    Game::APPLAB,
+    Game::GAMELAB,
+  ]
+
+  ##########################################################
+  #   Debug View                                           #
+  ##########################################################
+  def index
+    @key_value_pairs = DatablockStorageKvp.where(project_id: @project_id)
+    @records = DatablockStorageRecord.where(project_id: @project_id)
+    @tables = DatablockStorageTable.where(project_id: @project_id)
+    @library_manifest = DatablockStorageLibraryManifest.instance.library_manifest
+    @storage_backend = ProjectUseDatablockStorage.use_data_block_storage_for?(params[:channel_id]) ? "Datablock Storage" : "Firebase"
+  end
+
+  ##########################################################
+  #   Key-Value-Pair API                                   #
+  ##########################################################
+
+  def set_key_value
+    raise StudentFacingError, "The value is too large. The maximum allowable size is #{DatablockStorageKvp::MAX_VALUE_LENGTH} bytes" if params[:value].length > DatablockStorageKvp::MAX_VALUE_LENGTH
```

Comment:
Just a nit here: Do students know about bytes at this point? Do they know how to convert characters or numbers to bytes? (Should we just tell them to try a smaller value)

Reply:
Our goal was to be a "bug compatible" replacement for Firebase, and we match Firebase exactly here, both in byte size and (I believe?) in not having error messages translated. We should do that as a follow-on PR after we migrate (100% compatible the whole time) from Firebsae.

https://github.com/code-dot-org/code-dot-org/blob/106c169006320e410e0c77e23fe6bca4f9e080b9/apps/src/storage/firebaseStorage.js/#L188-L190